### DISCUSSION TITLE
Switching to OpenTofu

### DISCUSSION THREAD

**Discussion by gtirloni** [opentofu/opentofu#510] [General]
I want to start a discussion about how teams can switch to OpenTofu in a progressive way.

In the past, certain pre-1.0 versions of Terraform would taint the state file and not allow earlier versions to be used anymore.

Do we expect that simply running an `opentofu plan` command on an otherwise Terraform-managed environment would taint the state file in such a way that Hashicorp's Terraform would refuse to work afterwards? We can always restore the state file but I'm thinking from a least surprise perspective.

Are the other things that teams should worry about when planning a switch to OpenTofu?

**Comment by davidjeddy** [comment]
`taint the state file in such a way that Hashicorp's Terraform would refuse to work afterwards?`

I think this would make migrating a very scary one-way migration. When evaluating if a new tool is worth the effort to onboard the ability to not negatively impact existing configurations and resources is a high value consideration.

IMO OpenTofu should stay as close to binary and operationally compatible with Terraform for the foreseeable future. Like how MariaDB was a drop in binary replacement for MySQL for many years.

OR, worse case, an additional tool to convert the state file between OTF and TF.

**Reply by gtirloni** [reply]
Agreed. I meant the opposite: Hashicorp implements the logic to refuse states that were touched by OpenTofu. So OpenTofu would need to be stealth in this case, I hope that doesn't cause any issues with Hashicorp.

**Reply by jake-mok-nelson** [reply]
@gtirloni , where is this logic? I find this really surprising and wasn't able to find anything about opentf or opentofu in Terraform.

**Reply by gtirloni** [reply]
That's because this is conjecture.

**Comment by sftim** [comment]
We can use CI to validate that Terraform 1.0 to 1.5 (MPL) do not balk at opening a state file that OpenTofu has worked on. We can't tell what Hashicorp intend for later releases that aren't open source; however, validating compatibility with 1.0.x sounds like a useful thing to ensure.
